src/main/java/uk/gov/ons/ctp/response/sample/ingest/CsvIngesterBusiness.java

```diff
@@ -85,44 +105,68 @@ public CsvIngesterBusiness() {
     columnPositionMappingStrategy.setColumnMapping(COLUMNS);
   }
 
+  private String generateErrorReport(int maxErrors, List<BusinessSampleUnit> sampleUnitList, List<ValidationError> errors){
+    StringBuffer report = new StringBuffer();
+
+    report.append(String.format("timestamp: %s, valid samples: %d, errors: %d (fail at %d errors)\n",
+            new Date().toString(), sampleUnitList.size(), errors.size(), maxErrors));
+    report.append(errors
+        .stream()
+        .map(ve -> String.format("Line %d: %s -> %s", ve.getLineNumber(), ve.getErrorType().name(), ve.getErrorDetail()))
+        .collect(Collectors.joining("\n")));
+
+    return report.toString();
+  }
+
+  private void validateFilename(String filename) throws IngesterCTPException {
+    if (!filename.toLowerCase().endsWith(".csv")){
+      throw new IngesterCTPException(CTPException.Fault.VALIDATION_FAILED, SampleSummaryDTO.ErrorCode.NotCsv,
+              String.format("%s is not a valid CSV file (must have .csv extension)", filename));
+    }
+  }
+
   public SampleSummary ingest(final SampleSummary sampleSummary, final MultipartFile file)
       throws Exception {
 
-    CSVReader csvReader = new CSVReader(new InputStreamReader(file.getInputStream()), ':');
-    String[] nextLine;
-    List<BusinessSampleUnit> samplingUnitList = new ArrayList<>();
-    int lineNumber = 0;
-    Set<String> unitRefs = new HashSet<>();
+      validateFilename(file.getOriginalFilename());
 
-      while((nextLine = csvReader.readNext()) != null) {
+      int maxErrors = this.appConfig.getSampleIngest().getMaxErrors();
+      CSVReader csvReader = new CSVReader(new InputStreamReader(file.getInputStream()), ':');
+      String[] nextLine;
+      List<BusinessSampleUnit> samplingUnitList = new ArrayList<>();
+      int lineNumber = 0;
+      Set<String> unitRefs = new HashSet<>();
+      List<ValidationError> errors = new ArrayList<>();
+
+      while((nextLine = csvReader.readNext()) != null && errors.size() < maxErrors) {
         lineNumber++;
 
-        try {
-          BusinessSampleUnit businessSampleUnit = processLine(columnPositionMappingStrategy, nextLine);
-          Optional<String> namesOfInvalidColumns = validateLine(businessSampleUnit);
+        BusinessSampleUnit businessSampleUnit = processLine(columnPositionMappingStrategy, nextLine);
+        Optional<String> namesOfInvalidColumns = validateLine(businessSampleUnit);
 
-          // If a unit ref is already registered
+        if (namesOfInvalidColumns.isPresent() || unitRefs.contains(businessSampleUnit.getSampleUnitRef())) {
```

Comment:
Why are we checking the `namesOfInvalidColumns.isPresent()` and `unitRefs.contains(businessSampleUnit.getSampleUnitRef()` twice?

Reply:
Fixed